src/Caching/StackExchangeRedis/src/RedisCache.cs

```diff
@@ -30,26 +30,38 @@ public class RedisCache : IDistributedCache, IDisposable
         private const string DataKey = "data";
         private const long NotPresent = -1;
 
-        private volatile ConnectionMultiplexer _connection;
-        private IDatabase _cache;
+        private readonly IConnectionMultiplexer _connection;
+        private readonly IDatabase _cache;
         private bool _disposed;
 
-        private readonly RedisCacheOptions _options;
         private readonly string _instance;
 
-        private readonly SemaphoreSlim _connectionLock = new SemaphoreSlim(initialCount: 1, maxCount: 1);
-
         public RedisCache(IOptions<RedisCacheOptions> optionsAccessor)
         {
             if (optionsAccessor == null)
             {
                 throw new ArgumentNullException(nameof(optionsAccessor));
             }
 
-            _options = optionsAccessor.Value;
+            RedisCacheOptions options = optionsAccessor.Value;
+
+             // This allows partitioning a single backend cache for use with multiple apps/services.
+            _instance = options.InstanceName ?? string.Empty;
 
-            // This allows partitioning a single backend cache for use with multiple apps/services.
-            _instance = _options.InstanceName ?? string.Empty;
+            _connection = options.ConnectionMultiplexer;
+            if (_connection == null)
+            {
+                if (options.ConfigurationOptions != null)
+                {
+                    _connection = ConnectionMultiplexer.Connect(options.ConfigurationOptions);
```

Comment:
It looks like @Tratcher first did it this way in https://github.com/aspnet/Caching/commit/38dd3caf812728ccc0f8e99c6ad2aec064892255#diff-9ed70d9e570b33d5dc0bc574eb1d3c175d6d11397886ca242f646c6be97712dcR39, but then moved it out of the constructor during the rest of the initial implementation. See https://github.com/aspnet/Caching/pull/23/.

I assume there are a few problems with establishing the connection in the construtor:

1. What happens when it fails?
2. It could take a long time, when constructors are typically fast.
3. Not possible to be `async` any more.